clients/venice-client/src/main/java/com/linkedin/venice/fastclient/InternalAvroStoreClient.java

```diff
@@ -1,58 +1,197 @@
 import com.linkedin.venice.client.store.streaming.StreamingCallback;
+import com.linkedin.venice.client.store.streaming.VeniceResponseCompletableFuture;
 import com.linkedin.venice.client.store.streaming.VeniceResponseMap;
+import com.linkedin.venice.client.store.streaming.VeniceResponseMapImpl;
+import com.linkedin.venice.compute.ComputeRequestWrapper;
+import com.linkedin.venice.read.RequestType;
+import com.linkedin.venice.router.exception.VeniceKeyCountLimitException;
+import com.linkedin.venice.utils.concurrent.VeniceConcurrentHashMap;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
 
 
 /**
  * All the internal implementations of different tiers should extend this class.
  * This class adds in {@link RequestContext} object for the communication among different tiers.
  */
 
-public abstract class InternalAvroStoreClient<K, V> implements AvroGenericStoreClient<K, V> {
+public abstract class InternalAvroStoreClient<K, V> implements AvroGenericReadComputeStoreClient<K, V> {
+  public abstract ClientConfig getClientConfig();
+
+  @Override
+  public final boolean isProjectionFieldValidationEnabled() {
+    return getClientConfig().isProjectionFieldValidationEnabled();
+  }
+
+  @Override
   public CompletableFuture<V> get(K key) throws VeniceClientException {
-    return get(new GetRequestContext(), key);
+    return get(new GetRequestContext(false), key);
   }
 
   protected abstract CompletableFuture<V> get(GetRequestContext requestContext, K key) throws VeniceClientException;
 
+  @Override
+  public final CompletableFuture<Map<K, V>> batchGet(Set<K> keys) throws VeniceClientException {
+    // Since user has invoked batchGet directly, then we do not want to allow partial success
+    return batchGet(new BatchGetRequestContext<>(false), keys);
+  }
+
+  protected CompletableFuture<Map<K, V>> batchGet(BatchGetRequestContext<K, V> requestContext, Set<K> keys)
+      throws VeniceClientException {
+    return getClientConfig().useStreamingBatchGetAsDefault()
+        ? batchGetUsingStreamingBatchGet(requestContext, keys)
+        : batchGetUsingSingleGet(keys);
+  }
+
+  private CompletableFuture<Map<K, V>> batchGetUsingStreamingBatchGet(
+      BatchGetRequestContext<K, V> requestContext,
+      Set<K> keys) throws VeniceClientException {
+    CompletableFuture<Map<K, V>> resultFuture = new CompletableFuture<>();
+    CompletableFuture<VeniceResponseMap<K, V>> streamingResultFuture = streamingBatchGet(requestContext, keys);
+
+    streamingResultFuture.whenComplete((response, throwable) -> {
+      if (throwable != null) {
+        resultFuture.completeExceptionally(throwable);
+      } else {
+        resultFuture.complete(response);
```

Comment:
I believe I had it in this way before, but for some reason that I do not recall right now, I had to handle the partial response in `DispatchingAvroGenericStoreClient` and added the following comment:
```
            // TODO: It would be great to move this to InternalAvroStoreClient, but the callback is a
            // "StatTrackingCallback" and it needs the exceptions to be called out to not misrepresent in the stats.
```

I'll try again to see what breaks if I move that to `InternalAvroStoreClient`